plan/predicate_push_down.go

```diff
@@ -47,10 +49,229 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
 	return expr
 }
 
+// propagateConstant propagate constant values of equality predicates and inequality predicates in a condition.
+func propagateConstant(conditions []expression.Expression) []expression.Expression {
+	if len(conditions) == 0 {
+		return conditions
+	}
+	// Propagate constants in equality predicates.
+	// e.g. for condition: "a = b and b = c and c = a and a = 1";
+	// we propagate constant as the following step:
+	// first: "1 = b and b = c and c = 1 and a = 1";
+	// next:  "1 = b and 1 = c and c = 1 and a = 1";
+	// next:  "1 = b and 1 = c and 1 = 1 and a = 1";
+	// next:  "1 = b and 1 = c and a = 1";
+
+	// e.g for condition: "a = b and b = c and b = 2 and a = 1";
+	// we propagate constant as the following step:
+	// first: "a = 2 and 2 = c and b = 2 and a = 1";
+	// next:  "a = 2 and 2 = c and b = 2 and 2 = 1";
+	// next:  "0"
+	isHandled := make([]bool, len(conditions))
+	type transitiveEqualityPredicate map[string]*expression.Constant // transitive equality predicates between one column and one constant
+	equalities := make(transitiveEqualityPredicate, 0)
+	for i := 0; i < len(conditions); i++ {
+		if isHandled[i] {
+			continue
+		}
+		if len(equalities) != 0 {
+			conditions[i] = constantSubstitute(equalities, conditions[i])
+		}
+		expr, ok := conditions[i].(*expression.ScalarFunction)
+		if !ok {
+			continue
+		}
+		// process the included OR conditions recursively to do the same for CNF item.
+		switch expr.FuncName.L {
+		case ast.OrOr:
+			expressions := expression.SplitNormalFormItems(conditions[i], ast.OrOr)
+			newExpression := make([]expression.Expression, 0)
+			for _, v := range expressions {
+				newExpression = append(newExpression, propagateConstant([]expression.Expression{v})...)
+			}
+			conditions[i] = expression.ComposeConditionWithBinaryOp(newExpression, ast.OrOr)
+			isHandled[i] = true
+		case ast.AndAnd:
+			newExpression := propagateConstant(expression.SplitNormalFormItems(conditions[i], ast.AndAnd))
+			conditions[i] = expression.ComposeConditionWithBinaryOp(newExpression, ast.AndAnd)
+			isHandled[i] = true
+		case ast.EQ:
+			var (
+				col      *expression.Column
+				val      *expression.Constant
+				isColumn bool
+				bytes    []byte
+			)
+			left, ok1 := expr.Args[0].(*expression.Constant)
+			right, ok2 := expr.Args[1].(*expression.Constant)
+			if col, isColumn = expr.Args[0].(*expression.Column); ok2 && isColumn {
+				val = right
+			} else if col, isColumn = expr.Args[1].(*expression.Column); ok1 && isColumn {
+				val = left
+			} else {
+				continue
+			}
+			bytes, _ = codec.EncodeValue(bytes, types.NewStringDatum(col.FromID), types.NewIntDatum(int64(col.Position)))
+			equalities[string(bytes)] = val
+			isHandled[i] = true
+			i = -1
```

Comment:
e.g.  "a = b and a = 1 and b = c"
when find "a = 1", set all "a" to "1" in the condition except "a=1"
so here set i = -1 to restart the loop to propagate equality predicates